zypp/target/RpmPostTransCollector.cc

```diff
@@ -208,12 +335,59 @@ namespace zypp
           return _ptrTmpdir->path();
         }
 
+        /** Return whether RpmHeader has a  %posttrans. */
+        bool headerHasPosttrans( rpm::RpmHeader::constPtr pkg_r ) const
+        {
+          bool ret = false;
+          if ( pkg_r ) {
+            std::string prog( pkg_r->tag_posttransprog() );
+            if ( not prog.empty() && prog != "<lua>" )  // by now leave lua to rpm
+              ret = true;
+          }
+          return ret;
+        }
+
+        /** Cache RpmHeader for consecutive hasPosttransScript / collectScriptForPackage calls.
+         * @return RpmHeader::constPtr IFF \a rpmPackage_r has a %posttrans
+         */
+        rpm::RpmHeader::constPtr getHeaderIfPosttrans( const Pathname & rpmPackage_r )
+        {
+          if ( _headercache.first == rpmPackage_r )
+            return _headercache.second;
+
+          rpm::RpmHeader::constPtr ret { rpm::RpmHeader::readPackage( rpmPackage_r, rpm::RpmHeader::NOVERIFY ) };
+          if ( ret ) {
+            if ( not headerHasPosttrans( ret ) )
+              ret = nullptr;
+          } else {
+            WAR << "Unexpectedly this is no package: " << rpmPackage_r << endl;
+          }
+          _headercache = std::make_pair( rpmPackage_r, ret );
+          return ret;
+        }
+
+        /** Retrieve "dump_posttrans: install" lines from \a dumpfile_r and pass n,v,r,a to the \a consumer_r. */
+        void recallFromDumpfile( const Pathname & dumpfile_r, std::function<void(std::string,std::string,std::string,std::string)> consume_r )
```

Comment:
For a impl class thats only used in the cc file anyway we could use a templated callback instead of std::function.
`std::is_invocable<void(string,string,string,string)>` can be used as constraint and to document the callable we need.

Reply:
Yes we could.  Though the method is only called once and this only if rpm was downgraded within the transaction to a version not supporting --runposttrans. So the runtime penalty is probably neglectable.

Reply:
Right